[SE-0293] Update Library-Evolution Documentation

filip-sakel opened the pull request:
**This PR:**
* Fixes the link to [SE-0293](https://github.com/apple/swift-evolution/blob/main/proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md) in the Changelog; and
* Updates the [Library Evolution document](https://github.com/apple/swift/blob/main/docs/LibraryEvolution.rst) for SE-0293.

filip-sakel committed:
Add reference to SE-0293 in Changelog.

filip-sakel committed:
Update LibraryEvolution.rst for SE-0293.

xwu commented on docs/LibraryEvolution.rst:
```diff
@@ -413,8 +420,14 @@ stored subscripts. This means that the following changes are permitted:
 - Reordering generic requirements (but not the generic parameters themselves).
 - Adding a default argument expression to an index parameter.
+- Adding, changing, reordering, or removing property wrappers that either are 
+  implementation-detail or belong in a composition leading with an
```
Nit: Can you maybe clarify what "leading with" means in this context, maybe with a terse example? It makes sense to me fresh from reviewing a series of property wrapper-related proposals, but it may not make sense to a general audience (or even to me in a few months!).

hborla replied:
Agreed, I commonly use the phrase "outermost property wrapper" to describe the first property wrapper in a composition chain

filip-sakel replied:
I've changed `compositions leading with API wrappers` to `compositions where the outermost wrapper is API-level`. I've also added some brief examples of the discussed wrappers; let me know if I should change something.

filip-sakel committed:
Add property-wrapper examples.

hborla commented on docs/LibraryEvolution.rst:
```diff
@@ -145,14 +145,24 @@ The following changes are permitted:
+  the outermost wrapper is implementation-detail
+  (``@ImplementationDetail @MyWrapper``).
+- Changing or removing a default argument is a `binary-compatible source-breaking change`.
+- Adding, changing, or removing arguments in the property-wrapper custom attribute
```
Hmm I don't think this bullet point is right - adding or removing an argument in an API-level property wrapper attribute is an ABI and source breaking change. Changing the argument once it's there is fine.

filip-sakel replied:
Right. I was thinking that the behavior would be similar to defaulted parameters but didn’t consider that wrapper arguments are chosen by the function author.

hborla replied:
I think the way this is phrased in the proposal is fine. The section in the proposal that you mentioned specifies that this inference is per wrapper-application, not per wrapper type. I think the right way to phrase it here is to say that changing a property wrapper attribute from API to implementation-detail or vice versa is a source and ABI breaking change.

As far as arguments in composition chains - the check should be if any of the attributes have an argument. If the implementation doesn't do that then it's a bug.

filip-sakel replied:
> I think the way this is phrased in the proposal is fine. The section in the proposal that you mentioned specifies that this inference is per wrapper-application, not per wrapper type.

Agreed. I'm just worried that the wrapper-level API vs implementation-detail distinction may be more confusing with potential future exceptions.

> As far as arguments in composition chains - the check should be if any of the attributes have an argument. If the implementation doesn't do that then it's a bug.

That check guarantees that no wrapper attributes with arguments are allowed to be enclosed in a composition with an API-level outer wrapper, correct?

>  I think the right way to phrase it here is to say that changing a property wrapper attribute from API to implementation-detail or vice versa is a source and ABI breaking change.

Excellent suggestion; I simplified the rule and place it under the "not allowed" section:

---

An ABI-public function may not change an API-level property-wrapper attribute to an implementation-detail one and vice versa, if it is the only wrapper applied to a given parameter or the outermost wrapper in a composition.

hborla commented on docs/LibraryEvolution.rst:
```diff
@@ -413,8 +423,16 @@ stored subscripts. This means that the following changes are permitted:
+  (``@ImplementationDetail @MyWrapper``).
 - Changing or removing a default argument is a `binary-compatible
   source-breaking change`.
+- Adding, changing, or removing arguments in the property-wrapper custom attribute
```
Same comment as above

hborla commented on docs/LibraryEvolution.rst:
```diff
@@ -778,9 +796,17 @@ counterparts with a few small changes:
 - Adding a default argument expression to an index parameter is permitted.
 - Changing or removing a default argument is a `binary-compatible
   source-breaking change`.
+- Adding, changing, or removing arguments in the property-wrapper custom attribute
```
Same comment as above

filip-sakel committed:
Mark adding/removing PP arguments as ABI-breaking.

filip-sakel committed:
Typo fix.

filip-sakel committed:
Show that PP arguments change API significance.

filip-sakel committed:
Update examples for wrapper arguments.

xwu commented on docs/LibraryEvolution.rst:
```diff
@@ -145,14 +145,30 @@ The following changes are permitted:
-- Changing or removing a default argument is a `binary-compatible
-  source-breaking change`.
+- Adding, changing, reordering, or removing property wrappers that either are 
+  implementation-detail (``@ImplementationDetail``) or in a composition where
```
Sorry, what is this `@ImplementationDetail` wrapper? Is this a new/upcoming feature?

hborla replied:
I don't think that's a real wrapper, and I don't think including it in this bullet point adds any clarity.

filip-sakel replied:
OK, I'll remove it.

filip-sakel committed:
Simplify property-wrapper-attribute argument rules

filip-sakel committed:
Merge branch 'main' into filip-sakel-SE0293-gardening

xwu commented on docs/LibraryEvolution.rst:
```diff
@@ -145,14 +145,22 @@ The following changes are permitted:
 - An ABI-public function may not change its parameters or return type.
+- An ABI-public function may not, in any way, change API-level property 
+  wrappers (``@API``) or compositions where the outermost wrapper is 
+  API-level (``@API @MyWrapper``).
```
Nit: these still read as though there's some built-in feature named `@API` (or `@ImplementationDetail`); if I understand @hborla's point, I think all of these examples in parens can be deleted.

filip-sakel replied:
OK, I'll remove them. Are the bullet points clear without any examples, though, or would you like to see a different kind of example?

filip-sakel committed:
Remove unnecessary code  examples.

filip-sakel commented:
I removed the misleading code examples, per @xwu's review, and I think I've addressed all the feedback. So is it ready to merge or do you have any other suggestions?

xwu commented:
Thanks--LGTM!

hborla commented:
Thank you, @filip-sakel!

@swift-ci please smoke test
